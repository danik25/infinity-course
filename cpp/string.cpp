/****************************************************************
*                              String
* 						=====================
* File Name: string.cpp
* Related files: string.hpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#include <cstring> /* strlen */
#include <iostream> /* cout */
#include "string.hpp"

namespace hrd9
{
	/* constructor */
	String::String(const char* str)
	{
		size_t len = strlen(str);
		m_str = new char [len + 1];
		strcpy(m_str, str);
	}

	/* distructor */
	String::~String()
	{
		delete [](m_str);
	}

	/* copy constructor */
	String::String(const String& other) : m_str(init_new_str(other.m_str))
	{
		;
	}

	/* copy assignment operator */
	String& String :: operator=(const String& other)
	{
		delete[] m_str;
		m_str = init_new_str(other.m_str);

		return(*this);
	}

/*===================== non member functions =====================*/

/*===================== member functions =====================*/

	String& String::operator+=(const String& other)
	{
		size_t total_len = strlen(m_str) + strlen(other.m_str);

		String tmp = m_str;
		delete [] m_str;

		m_str = new char [total_len + 1];
		char * runner = m_str;

		strcpy(runner, tmp.m_str);
		strcat(runner, other.m_str);

		return(*this);
	}

	size_t String::length() const
	{
		return(strlen(m_str));
	}

	const char& String::operator[](size_t index) const
	{
		return(m_str[index]);
	}

	char& String::operator[](size_t index)
	{
		return(m_str[index]);
	}

	const char* String::to_c_str() const
	{
		return(m_str);
	}

/*===================== friend functions =====================*/

	const String operator+(const String& lhs, const String& rhs)
	{
		String ans = lhs;
		ans += rhs;

		return(ans);
	}

	bool operator==(const String& lhs, const String& rhs)
	{
		return(!strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator!=(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator>(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator<(const String& lhs, const String& rhs)
	{
		return(strcmp(rhs.m_str, lhs.m_str));
	}

	std::ostream& operator<<(std::ostream& os, const String& str)
	{
		os << str.m_str;
		return(os);
	}

/*===================== assistance functions =====================*/

	char* String::init_new_str(const char *str)
	{
		size_t len = strlen(str);
		char *new_str = new char [len + 1];
		strcpy(new_str, str);

		return(new_str);
	}
}