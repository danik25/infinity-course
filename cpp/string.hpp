/****************************************************************
*                              String
* 						=====================
* File Name: string.hpp
* Related files: string.cpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#ifndef ILHRD_STRING_HPP
#define ILHRD_STRING_HPP

#include <iosfwd> //ostream forward declarations

namespace hrd9
{

class String;
//non member functions:
const String operator+(const String& lhs, const String& rhs);
bool operator==(const String& lhs, const String& rhs);
bool operator!=(const String& lhs, const String& rhs);
bool operator>(const String& lhs, const String& rhs);
bool operator<(const String& lhs, const String& rhs);
std::ostream& operator<<(std::ostream& os, const String& str);
std::istream& operator>>(std::istream& is, String& str);

class String
{

public:
	// explicitly non explicit constructor
	String(const char* str = "");
	String(const String& other);
	String& operator=(const String& other);
	~String();

	String& operator+=(const String& other);
	size_t length() const;

	const char& operator[](size_t index) const; //read string[]
	char& operator[](size_t index);             //string[i] assignment

	const char* to_c_str() const;

private:
	friend bool operator==(const String& lhs, const String& rhs);
	friend bool operator!=(const String& lhs, const String& rhs);
	friend bool operator>(const String& lhs, const String& rhs);
	friend bool operator<(const String& lhs, const String& rhs);
	friend std::ostream& operator<<(std::ostream& os, const String& str);
	friend std::istream& operator>>(std::istream& is, String& str);

	char* init_new_str(const char *str);

	char* m_str;
};
} //namespace hrd9




#endif /* ILHRD_STRING_HPP */