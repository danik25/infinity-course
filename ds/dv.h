/****************************************************************
*                       Data Structures - Dynamic vector        *
* Related files: dv.c, dv_test.c, dv.h                          *
* #Version: V.01                                                *
* writer: Dani Benjmain                                         *
* reviewer: Kobi Madrish                                        *
* last update: 1.4.19                                           *
****************************************************************/ 
/*
*   Dynamic vector is a sequence container representing 
*   array that can change in size.
*   Dynamic vector use contiguous storage locations for its elements,
*   and its size can change dynamically.
*/  

#ifndef _DV_H_
#define _DV_H_

#include <stdio.h> /* size_t */

typedef struct dynamic_vector dv_t;

/*  
*   dv_t *DVCreate(size_t capacity, size_t element_size)  
* 
*   Description:
*       Create a dynamic vector, initializing its contents 
*       depending on given capacity and element size. 
*
*   Params:
*       capacity:  the maximum size of the vector. [size_t]
*       element_size: size of element. [size_t]  
*    
*   Return value: 
*       the new vector. [dv_t *]
*/
dv_t *DVCreate(size_t capacity, size_t element_size);


/*  
*   void DVDestroy(dv_t *vector)
*   
*   Description:
*       Destroy the vector, and deallocate all the storage 
*       capacity allocated by the vector using its allocator.  
*
*   Params:
*       vector: the vector to destroy. [dv_t *] 
*/
void DVDestroy(dv_t *vector); 


/*  
*   int DVPushBack(dv_t *vector, const void *element) 
*  
*   Description:
*       Adds a new element at the end of the vector, 
*       after its current last element.
*    
*   Params:
*       vector: the vector to append to it. [dv_t *]
*       element: the element to be add. [Const void *]
*      
*   Return value: 
*       0 if success, 1 if fail (if the allocation 
*       request does not succeed). [int]
*/
int DVPushBack(dv_t *vector, const void *element);


/*  
*   void DVPopBack(dv_t *vector)  
* 
*   Description:
*       Removes the last element in the vector. 
*  
*   Params:
*       vector: the vector to remove from. [dv_t *]
*/
void DVPopBack(dv_t *vector);


/*  
*   void *DVGetItemAddress(const dv_t *vector, size_t index)  
*
*   NOTE: 
*       DVPushBack() and DVPopBack invalidates 
*       DVGetItemAddress()'s reurn value !!!!!
*
*   Description:
*       Get the address of the item at 
*       position <index> in the vector.
*        
*   Params:
*       vector: the vector to get the item address from it. [const dv_t *]
*       index: position of an element in the vector. [size_t]    
*  
*   Return value: 
*       the address of the item at the specified position in the vector [void *]
*/
void *DVGetItemAddress(const dv_t *vector, size_t index);


/*  
*   size_t DVSize(const dv_t *vector)
*  
*   Description:
*       Get the number of elements in the vector.
*    
*   Params:
*       vector: the vector to get its size. [const dv_t *]
*      
*   Return value: 
*       the nubmer of elements in the vector. [size_t]
*/
size_t DVSize(const dv_t *vector);


/*  
*   size_t DVCapacity(const dv_t *vector)
*  
*   Description:
*       Get size of current allocated storage capacity.
*    
*   Params:
*       vector: the vector to get its capacity. [const dv_t *]
*      
*   Return value: 
*       the size of the storage space currently allocated for 
*       the vector, expressed in terms of elements. [size_t]
*/
size_t DVCapacity(const dv_t *vector);
 
 
/*  
*   int DVReserve(dv_t *vector, size_t new_capacity)
*  
*   Description:
*       Request a change in capacity of the vector.
*
*   NOTE: If new_capacity < size, do not make any changes and return success.
*    
*   Params:
*       vector: the vector to change its capacity. [const dv_t *]
*       new_capacity: the new capacity to be set.[size_t]
*      
*   Return value: 
*       0 if success, 1 if fail(if the allocation 
*       request does not succeed). [int] 
*/
int DVReserve(dv_t *vector, size_t new_capacity);

#endif /* _DV_H_ */
