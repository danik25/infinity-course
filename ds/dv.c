#include <stdlib.h> /*malloc*/
#include <assert.h> /*assert*/
#include <string.h> /*memmove*/
#include "dv.h"

struct dynamic_vector{

	void *array;
	size_t size; /* number of elements */ 
	size_t init_capacity; /* minimum capacity */
	size_t current_capacity; 
	size_t element_size;

};

dv_t *DVCreate(size_t capacity, size_t element_size)
{
    dv_t *vector = (dv_t*) malloc(sizeof(dv_t));
    vector -> array = (dv_t*) malloc(capacity * element_size);
    vector -> element_size = element_size;
    vector -> init_capacity = capacity;
    vector -> current_capacity = capacity;
    vector -> size = 0;
    
    if (NULL == vector || NULL == vector -> array)
    {
    	return NULL;
    }
    return vector;
}

void DVDestroy(dv_t *vector)
{
    assert(vector);
	
    free(vector -> array);
    vector -> array = NULL;
    free(vector);
    vector = NULL;
}

int DVPushBack(dv_t *vector, const void *element)
{
    dv_t *tmp = NULL;
    assert(vector);
	assert(element);
	if (vector -> size == vector -> current_capacity )
	{
	    tmp = realloc(vector -> array,
	                 2 * (vector -> current_capacity * vector -> element_size));
	    if (!tmp)
	    {
	        return 1;
	    }
	    vector -> array = tmp;
	    vector -> current_capacity *= 2;
	}

	memmove((char*)vector -> array + (vector -> size * vector -> element_size),
	    element, vector -> element_size);
	vector -> size += 1;
		
	return 0;
}

void *DVGetItemAddress(const dv_t *vector, size_t index)
{
    assert(vector);
    if(index >= vector -> size)
    {
        return (NULL);
    }
    return ((char*)vector -> array + index * (vector -> element_size));
}

void DVPopBack(dv_t *vector)
{
    dv_t *tmp = NULL;
    size_t tmp_vector = vector -> current_capacity;
    assert(vector);
    
    if(0 == vector -> size)
    {
        return;
    }  
    
    vector -> size -= 1;
    /* making the vector samller by half */
    if (vector -> size * 4 <= vector -> current_capacity)
    {
        if(vector -> init_capacity * 2 <= vector -> current_capacity)
        {
            tmp_vector /= 2;
        }
        else
        {
            tmp_vector = vector -> init_capacity;
        }
        
        tmp = realloc(vector -> array, 
                            tmp_vector * vector -> element_size);
        /* checking if the reallocation failed */
        if (tmp)
        {
            vector -> current_capacity = tmp_vector;
            vector -> array = tmp;
        }
    }
}

size_t DVSize(const dv_t *vector)
{
    assert(vector);
    return vector -> size;
}

size_t DVCapacity(const dv_t *vector)
{
    assert(vector);
    return vector -> current_capacity;
}

int DVReserve(dv_t *vector, size_t new_capacity)
{
    dv_t *tmp = NULL;
    assert(vector);
    if (new_capacity > vector -> current_capacity)
    {
        tmp = realloc(vector -> array, 
                        new_capacity * vector -> element_size);
        /* checking if the reallocation failed */
        if (tmp)
        {
            vector -> array = tmp;
            vector -> current_capacity = new_capacity;
            return 0;
        }
    }
    return 1;
}



