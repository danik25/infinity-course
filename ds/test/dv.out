/home/xyz/git/dani-benjamin/ds/dv.out